File: 03_binary_search/03_find_minimum_in_rotated_sorted_array/find_minimum_in_rotated_sorted_array.cpp
#include <vector>

class Solution
{
  public:
    int findMin(std::vector<int>& nums)
    {
        if (nums.front() < nums.back() || nums.size() == 1)
            return nums.front();
        if (nums.size() == 2)
            return std::min(nums[0], nums[1]);

        size_t l = 0;
        size_t r = nums.size() - 1;
        size_t m = avg(l, r);
        while (1)
        {
            if (l + 1 == m && nums[l] > nums[m]) // Minimun is found
                return nums[m];
            else if (m + 1 == r && nums[m] > nums[r])
                return nums[r];

            if (nums[l] < nums[m]) // minimum is to the right of middle
                l = m;
            else // nums[l] > nums[m] // minimu is to the left of middle
                r = m;
            m = avg(l, r);
        }
    }

  private:
    int avg(int a, int b)
    {
        return (a + b) / 2;
    }
};

#include <iostream>

int main()
{
    Solution         solution;
    std::vector<int> a {0, 1, 2};
    std::vector<int> b {2, 0, 1};
    std::vector<int> c {1, 2, 0};
    std::cout << solution.findMin(a) << "\n";
    std::cout << solution.findMin(b) << "\n";
    std::cout << solution.findMin(c) << "\n";
}

File: NeetCode/03_binary_search/03_find_minimum_in_rotated_sorted_array/find_minimum_in_rotated_sorted_array.py
from typing import List


class Solution:
    def avg(self, a: int, b: int):
        return (a + b) // 2

    def findMin(self, nums: List[int]) -> int:
        if len(nums) == 1 or nums[0] < nums[-1]:
            return nums[0]
        if len(nums) == 2:
            return min(nums[0], nums[1])
        l: int = 0
        r: int = len(nums) - 1
        m: int = self.avg(l, r)
        while True:
            if l + 1 == m and nums[l] > nums[m]:
                return nums[m]
            elif m + 1 == r and nums[m] > nums[r]:
                return nums[r]

            if nums[l] < nums[m]:
                l = m
            else:
                r = m
            m = self.avg(l, r)
